nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/upload/RUploadHandler.java

```diff
@@ -0,0 +1,114 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.sonatype.nexus.repository.Repository;
+import org.sonatype.nexus.repository.r.internal.RFormat;
+import org.sonatype.nexus.repository.r.RHostedFacet;
+import org.sonatype.nexus.repository.rest.UploadDefinitionExtension;
+import org.sonatype.nexus.repository.security.ContentPermissionChecker;
+import org.sonatype.nexus.repository.security.VariableResolverAdapter;
+import org.sonatype.nexus.repository.storage.Asset;
+import org.sonatype.nexus.repository.storage.StorageFacet;
+import org.sonatype.nexus.repository.upload.AssetUpload;
+import org.sonatype.nexus.repository.upload.ComponentUpload;
+import org.sonatype.nexus.repository.upload.UploadDefinition;
+import org.sonatype.nexus.repository.upload.UploadFieldDefinition;
+import org.sonatype.nexus.repository.upload.UploadFieldDefinition.Type;
+import org.sonatype.nexus.repository.upload.UploadHandlerSupport;
+import org.sonatype.nexus.repository.upload.UploadResponse;
+import org.sonatype.nexus.repository.view.PartPayload;
+import org.sonatype.nexus.transaction.UnitOfWork;
+
+import static org.sonatype.nexus.repository.r.internal.RPathUtils.path;
+import static org.sonatype.nexus.repository.r.internal.RPathUtils.removeInitialSlashFromPath;
+
+/**
+ * @since 1.1.0
+ */
+@Singleton
+@Named(RFormat.NAME)
+public class RUploadHandler
+    extends UploadHandlerSupport
+{
+  private final VariableResolverAdapter variableResolverAdapter;
+
+  private final ContentPermissionChecker contentPermissionChecker;
+
+  private UploadDefinition definition;
+
+  final static String PATH_ID = "pathId";
+
+  final static String PACKAGE_PATH_DISPLAY = "Package Path";
+
+  @Inject
+  public RUploadHandler(@Named("simple") final VariableResolverAdapter variableResolverAdapter,
+                        final ContentPermissionChecker contentPermissionChecker,
+                        final Set<UploadDefinitionExtension> uploadDefinitionExtensions)
+  {
+    super(uploadDefinitionExtensions);
+    this.variableResolverAdapter = variableResolverAdapter;
+    this.contentPermissionChecker = contentPermissionChecker;
+  }
+
+  @Override
+  public UploadResponse handle(final Repository repository, final ComponentUpload upload) throws IOException {
+    final AssetUpload assetUpload = upload.getAssetUploads().get(0);
+    final PartPayload payload = assetUpload.getPayload();
+    final Map<String, String> fields = assetUpload.getFields();
+    final String uploadPath = removeInitialSlashFromPath(fields.get(PATH_ID));
+    final String assetPath = path(uploadPath, payload.getName());
+
+    ensurePermitted(repository.getName(), RFormat.NAME, assetPath, Collections.emptyMap());
+
+    try {
+      UnitOfWork.begin(repository.facet(StorageFacet.class).txSupplier());
+      Asset asset = repository.facet(RHostedFacet.class).upload(assetPath, payload);
```

Comment:
Question: What happens if redeploy is disabled and an already existing package is uploaded? Does this return a correct `UploadResponse` i.e. unable to upload?

Reply:
here is the message "Repository does not allow updating assets: r-hosted"